dotnet-authserver/tests/TeacherIdentity.AuthServer.EndToEndTests/SignOut.cs

```diff
@@ -9,53 +9,96 @@ public class SignOut : IClassFixture<HostFixture>
     [Fact]
-    public async Task User_CanSignOut()
+    public async Task SignOutFromClient()
     {
         var user = await _hostFixture.TestData.CreateUser(userType: UserType.Default);
 
         await using var context = await _hostFixture.CreateBrowserContext();
         var page = await context.NewPageAsync();
 
-        // Start on the client app and try to access a protected area
+        await page.StartOAuthJourney();
 
-        await page.GotoAsync("/");
-        await page.ClickAsync("text=Profile");
+        await page.SignInFromLandingPage();
 
-        // Fill in the sign in form (email + PIN)
+        await page.SubmitEmailPage(user.EmailAddress);
 
-        await page.FillAsync("text=Enter your email address", user.EmailAddress);
-        await page.ClickAsync("button:text-is('Continue')");
+        await page.SubmitEmailConfirmationPage(_hostFixture.CapturedEmailConfirmationPins.Last().Pin);
 
-        var pin = _hostFixture.CapturedEmailConfirmationPins.Last().Pin;
-        await page.FillAsync("text=Enter your code", pin);
-        await page.ClickAsync("button:text-is('Continue')");
+        await page.SubmitCompletePageForExistingUser();
 
-        // Should now be on the confirmation page
+        await page.AssertSignedInOnTestClient(user);
 
-        Assert.Equal(1, await page.Locator("data-testid=known-user-content").CountAsync());
-        await page.ClickAsync("button:text-is('Continue')");
+        await page.SignOutFromTestClient();
 
-        // Should now be back at the client, signed in
+        await page.AssertOnTestClient();
 
-        var clientAppHost = new Uri(HostFixture.ClientBaseUrl).Host;
-        var pageUrlHost = new Uri(page.Url).Host;
-        Assert.Equal(clientAppHost, pageUrlHost);
+        _hostFixture.EventObserver.AssertEventsSaved(
+            e => Assert.IsType<Events.UserSignedInEvent>(e),
+            e =>
+            {
+                var userSignedOut = Assert.IsType<Events.UserSignedOutEvent>(e);
+                Assert.Equal(DateTime.UtcNow, userSignedOut.CreatedUtc, TimeSpan.FromSeconds(10));
+                Assert.Equal(user.UserId, userSignedOut.User.UserId);
+                Assert.Equal(_hostFixture.TestClientId, userSignedOut.ClientId);
+            });
+    }
+
+    [Fact]
+    public async Task SignOutFromIdDirectly()
+    {
+        var user = await _hostFixture.TestData.CreateUser(userType: UserType.Default);
+
+        await using var context = await _hostFixture.CreateBrowserContext();
+        var page = await context.NewPageAsync();
+
+        await page.GoToAccountPage();
+
+        await page.SignInFromLandingPage();
+
+        await page.SubmitEmailPage(user.EmailAddress);
+
+        await page.SubmitEmailConfirmationPage(_hostFixture.CapturedEmailConfirmationPins.Last().Pin);
+
+        await page.SignOutFromAccountPageWithoutClientContext();
+
+        _hostFixture.EventObserver.AssertEventsSaved(
+            //e => Assert.IsType<Events.UserSignedInEvent>(e),
```

Comment:
There's a bug currently that prevents a `UserSignedInEvent` from being issued in this situation. I'll fix that separately.

Reply:
Is there a better way of testing if the user is signed in/out than using the Events?

Reply:
We could put something into the page HTML, I guess. These event checks are still useful, though (we wouldn't have caught this bug without them.)

Reply:
Fair point. I've added a marker `<span>` to TestClient that indicates logged in-ness and added a check on that to tests.

Reply:
looks good to me!